Add interface for L3 routing HW offload control via ip_forward flag

firscity opened the pull request.

firscity committed:
net: ipv4: add netevent for ip_forward flag update

Some network drivers implements FIB notifier interface for L3 routing
offload. In this case routes that are offloaded to hardware will not
be disabled when system routing will be turned off (ip_forward == 0).
The reason of this is absence of mechanism for drivers notification.

This commit adds netevent notification for sysctl.ipv4.ip_forward
changes. They can be received by kernel drivers via registered
netevent notifier. NETEVENT_IPV4_FORWARD_UPDATE will be sent everytime
after ip_forward change and driver will be able to manage offloaded
routes functionalities.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

firscity committed:
net: ethernet: renesas: add L3 offload handling via ip_forward flag

This commit introduces control for L3 routing HW offload using system
ip_forward flag (/proc/sys/net/ipv4/ip_forward or net.ipv4.ip_forward
sysctl). It implemented with netevent notifier mechanism and needed
to disable/enable HW forwarding from userspace.

Previously all routes that was added to HW was always active and
can not be disabled. Now HW forwarding can be controlled in the
same way as software one.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

LKomaryanskiy commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -4107,9 +4241,15 @@ static int renesas_eth_sw_remove(struct platform_device *pdev)
+
 	unregister_netdevice_notifier(&vlan_notifier_block);
 	unregister_fib_notifier(&init_net, &priv->fib_nb);
+	register_netevent_notifier(&netevent_notifier);
```
Seems like, there has to be a unregister, instead of register.

LKomaryanskiy commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -3982,6 +4051,65 @@ static struct notifier_block vlan_notifier_block __read_mostly = {
+	struct work_struct *work;
+
+	if (event != NETEVENT_IPV4_FORWARD_UPDATE)
+		return 0;
```
NIT: it can be changed to NOTIFY_DONE macro (it's equal to zero, so it's just a cosmetic change)

LKomaryanskiy commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -3982,6 +4051,65 @@ static struct notifier_block vlan_notifier_block __read_mostly = {
+
+	queue_work(priv->rswitch_netevent_wq, work);
+
+	return 0;
```
The same

LKomaryanskiy commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -3982,6 +4051,65 @@ static struct notifier_block vlan_notifier_block __read_mostly = {
+		return 0;
+
+	rn = net_generic(&init_net, rswitch_net_id);
+	priv = rn->priv;
```
NIT: You can use "glob_priv" variable to get rswitch_private structure.

LKomaryanskiy commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -3628,29 +3643,41 @@ void rswitch_add_ipv4_forward_all_types(struct l3_ipv4_fwd_param *param, struct
+
+	mutex_lock(&priv->ipv4_forward_lock);
+
+	list_add(&param_list[0].list, &routing_list->param_list);
```
If you move list_add section before hw setup, you have to add removing them in case of any hw error too.

LKomaryanskiy commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -3664,6 +3691,7 @@ static struct rswitch_ipv4_route *rswitch_get_route(struct rswitch_private *priv
 	bool default_present = false;
 
 	mutex_lock(&priv->rdev_list_lock);
+	mutex_lock(&priv->ipv4_forward_lock);
```
It seems like this lock can be redundant because forward flag status has to be kept in mind only when you setup hw.

LKomaryanskiy commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -2709,7 +2716,9 @@ static void rswitch_fib_event_add(struct rswitch_fib_event_work *fib_work)
 	new_routing_list->dev = dev;
 	INIT_LIST_HEAD(&new_routing_list->param_list);
 
+	mutex_lock(&dev->priv->ipv4_forward_lock);
```
It seems like this lock can be redundant because forward flag status has to be kept in mind only when you setup hw.

LKomaryanskiy commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -2741,6 +2750,7 @@ static void rswitch_fib_event_remove(struct rswitch_fib_event_work *fib_work)
 	if (!dev)
 		return;
 
+	mutex_lock(&dev->priv->ipv4_forward_lock);
```
It seems like this lock can be redundant because forward flag status has to be kept in mind only when you setup hw.

LKomaryanskiy commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -2035,31 +2043,28 @@ static int rswitch_stop(struct net_device *ndev)
+	struct l3_ipv4_fwd_param_list *l3_param_list;
+
+	mutex_lock(&priv->rdev_list_lock);
+	mutex_lock(&priv->ipv4_forward_lock);
```
It seems like this lock can be redundant because forward flag status has to be kept in mind only when you setup hw.

Deedone commented:
LGTM with Leonid's suggestions

Deedone commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -2035,31 +2044,27 @@ static int rswitch_stop(struct net_device *ndev)
+			list_for_each_entry(l3_param_list, &routing_list->param_list, list) {
+				if (l3_param_list->param->src_ip == src_ip &&
+				    l3_param_list->param->dst_ip == dst_ip) {
+					mutex_unlock(&priv->ipv4_forward_lock);
```
Missing lock

Deedone commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -3687,24 +3713,28 @@ static struct rswitch_ipv4_route *rswitch_get_route(struct rswitch_private *priv
+
+	mutex_lock(&fwd_work->priv->ipv4_forward_lock);
+	if (is_l3_exist(fwd_work->priv, fwd_work->src_ip, fwd_work->dst_ip))
 		return;
```
fwd_work leaks on early return

firscity committed:
net: renesas: rswitch: fix memory allocation for l3 params

This commit fixes issue with kcalloc() allocation of l3_params for
forwarding rule. Each rule contains 3 frame types, that were released
separately. But kcalloc-allocated memory should be released only once.
This caused strange malloc-related issues and kernel oops messages.

Also there was overallocation - 3 params (instead of 1) were allocated
for 1 frame type. This bug was also fixed during alloc flow re-working.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

LKomaryanskiy commented:
Good job!
Reviewed-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
